sdk/lib/rtl/rangelist.c

```diff
@@ -13,546 +13,1557 @@
 #define NDEBUG
 #include <debug.h>
 
+/* GLOBALS ******************************************************************/
+
+//extern PAGED_LOOKASIDE_LIST RtlpRangeListEntryLookasideList;
+
 /* TYPES ********************************************************************/
 
-typedef struct _RTL_RANGE_ENTRY
-{
-    LIST_ENTRY Entry;
-    RTL_RANGE Range;
-} RTL_RANGE_ENTRY, *PRTL_RANGE_ENTRY;
+/* RTLP_RANGE_LIST_ENTRY == RTL_RANGE + ListEntry */
+
+/*  FIXME ? ntddk.h */
+/* Flags for public functions: RtlFindRange(), RtlIsRangeAvailable() */
+#define RTL_RANGE_LIST_SHARED_OK           1
+#define RTL_RANGE_LIST_NULL_CONFLICT_OK    2
+/* ? ntddk.h */
+
+/* RTLP_RANGE_LIST_ENTRY.PrivateFlags */
+#define RTLP_ENTRY_IS_MERGED  1
 
 /* FUNCTIONS ***************************************************************/
 
-/**********************************************************************
- * NAME							EXPORTED
- * 	RtlAddRange
+static
+PRTLP_RANGE_LIST_ENTRY
+RtlpEntryFromLink(PVOID Link)
+{
+    return CONTAINING_RECORD(Link, RTLP_RANGE_LIST_ENTRY, ListEntry);
+}
+
+static
+BOOLEAN
+IsRangesIntersection(
+    _In_ PRTLP_RANGE_LIST_ENTRY Entry1,
+    _In_ PRTLP_RANGE_LIST_ENTRY Entry2)
+{
+    /* Not intersection when:
+       (Entry1->Start - Entry1->End) ... (Entry2->Start -
+       (Entry2->Start - Entry2->End) ... (Entry1->Start -
+    */
+
+    if (((Entry2->Start > Entry1->Start && Entry2->Start > Entry1->End) ||
+         (Entry1->Start > Entry2->Start && Entry1->Start > Entry2->End)))
+    {
+        /* No intersection */
+        return FALSE;
+    }
+    else
+    {
+        /* Entry1 and Entry2 overlap */
+        return TRUE;
+    }
+}
+
+CODE_SEG("PAGE")
+static
+VOID
+RtlpDeleteRangeListEntry(
+    _In_ PRTLP_RANGE_LIST_ENTRY DelEntry)
+{
+    PRTLP_RANGE_LIST_ENTRY RtlEntry;
+    PRTLP_RANGE_LIST_ENTRY NextRtlEntry;
+
+    PAGED_CODE_RTL();
+    DPRINT("RtlpDeleteRangeListEntry: DelEntry %p\n", DelEntry);
+
+    if (!(DelEntry->PrivateFlags & RTLP_ENTRY_IS_MERGED))
+    {
+        goto Finish;
+    }
+
+    RtlEntry = RtlpEntryFromLink(DelEntry->Merged.ListHead.Flink);
+    NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
+
+    while (&RtlEntry->ListEntry != &DelEntry->Merged.ListHead)
+    {
+        DPRINT("RtlpDeleteRangeListEntry: Free RtlEntry %p, [%I64X-%I64X]\n", RtlEntry, RtlEntry->Start, RtlEntry->End);
+
+        //ExFreeToPagedLookasideList(&RtlpRangeListEntryLookasideList, RtlEntry);
```

Comment:
I started learning and developing PnP (including rangelist) from the beginning of 2018 and it has been more than a year. As you can see, a lot of time has passed and .. the details are forgotten.
RtlpRangeListEntryLookasideList is used by Windows. I made a kind of "bookmarks in the book" so that later I would not spend a lot of time looking for these places when needed.
I can remove it if it bothers. But as usual, removing takes much less time than adding.